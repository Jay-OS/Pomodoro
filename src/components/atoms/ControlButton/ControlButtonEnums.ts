export enum controlButtonType {
    PRIMARY = 'controlPrimary',
    SECONDARY = 'controlSecondary',
    REVERSED = 'controlReversed',
}

export enum controlButtonSize {
    LARGE = 'controlLarge',
    MEDIUM = 'controlMedium',
    SMALL = 'controlSmall',
}
