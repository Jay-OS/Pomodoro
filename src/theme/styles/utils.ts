import { styleMapProperties } from './objectMaps';

const utils: styleMapProperties = {
    CENTER_ELEMENT: {
        position: 'absolute',
        top: '50%',
        left: '50%',
        transform: 'translate(-50%, -50%)',
    },
    CONTAINER_STYLES: {
        borderRadius: '10px',
        padding: '1rem',
    },
    NO_WRAP: {
        whiteSpace: 'nowrap',
    },
    HIDDEN: {
        position: 'absolute',
        top: '-9999px',
        left: '-9999px',
    },
    CONTAINER_OVERFLOW: {
        overflowX: 'hidden',
        overfloyY: 'auto',
    },
};

export default utils;
