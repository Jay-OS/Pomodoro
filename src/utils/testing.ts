import React from 'react';
import ReactDOM from 'react-dom';
import { act, render } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import { StyleSheetTestUtils } from 'aphrodite/no-important';

export const accessibilityChecks = (
    component: React.ReactElement,
    promise?: Promise<any>,
    rules = {}
) => {
    const axeOptions = { rules: { ...rules } };
    expect.extend(toHaveNoViolations);
    it('Has no accessibility issues', async () => {
        const { container } = render(component);
        if (promise) await act(() => promise);
        const results = await axe(container, axeOptions);
        expect(results).toHaveNoViolations();
    });
};

export const mountTest = (component: React.ReactElement, promise?: Promise<any>) => {
    it('Renders without crashing', async () => {
        if (promise) await act(() => promise);
        const div = document.createElement('div');
        div.setAttribute('id', 'root');
        ReactDOM.render(component, div);
    });
};

/**
 * ONLY TO BE USED TO ALLOW DEBUGGING OF INDIVIDUAL TESTS 
 * AND SHOULD NOT BE LEFT IN FINISHED UNIT TESTING CODE
 * AS IT WILL INFLUENCE THE OUTCOME OF ACCESSIBILITY TESTS
 */
export const disableAphrodite = () => {
    beforeEach(() => {
        StyleSheetTestUtils.suppressStyleInjection();
    });
    
    afterEach(() => {
        StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
    });
};
