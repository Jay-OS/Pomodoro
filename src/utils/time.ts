export const minutesToHoursAndMinutes = (minutes: number) => {
    const hoursAndMinutes = {
        hours: Math.floor(minutes / 60),
        minutes: minutes % 60,
    };

    return hoursAndMinutes;
};

export const minutesToHoursAndMinutesString = (minutes: number) => {
    const hoursAndMinutes = minutesToHoursAndMinutes(minutes);
    return `${hoursAndMinutes.hours}h ${hoursAndMinutes.minutes}m`;
};

export const minutesToMilliseconds = (minutes: number) => {
    return minutes * 60 * 1000;
};

export const millisecondsToMinutesAndSeconds = (milliseconds: number) => {
    const totalSeconds = millisecondsToSeconds(milliseconds);
    return {
        minutes: parseInt((totalSeconds / 60).toString()),
        seconds: totalSeconds % 60,
    };
};

export const millisecondsToMinutesAndSecondsString = (milliseconds: number) => {
    const minutesAndSeconds = millisecondsToMinutesAndSeconds(milliseconds);
    return {
        minutes: toTwoDigitString(minutesAndSeconds.minutes),
        seconds: toTwoDigitString(minutesAndSeconds.seconds),
    };
};

const millisecondsToSeconds = (milliseconds: number) => {
    return parseInt((milliseconds / 1000).toString());
};

const toTwoDigitString = (number: number) => {
    return number.toLocaleString('en-GB', {
        minimumIntegerDigits: 2,
        useGrouping: false,
    });
};
