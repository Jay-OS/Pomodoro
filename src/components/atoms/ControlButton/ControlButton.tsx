import * as React from 'react';
import { css } from 'aphrodite/no-important';

import { controlButtonType, controlButtonSize } from './ControlButtonEnums';

import controlButtonStyles from './ControlButtonStyles';

export interface ControlButtonProps {
    id?: string;
    title?: string;
    onClick: React.MouseEventHandler<HTMLButtonElement>;
    disabled?: boolean;
    ['aria-label']: string;
    buttonType: controlButtonType;
    buttonSize: controlButtonSize;
};

const ControlButton: React.FC<ControlButtonProps> = ({
    id,
    title,
    onClick,
    disabled,
    'aria-label': ariaLabel,
    buttonType,
    buttonSize,
    children
}) => (
    <button
        id={id}
        title={title}
        className={css(controlButtonStyles.controlButton, controlButtonStyles[buttonType], controlButtonStyles[buttonSize])}
        disabled={disabled}
        onClick={onClick}
        aria-label={ariaLabel}
    >
        {children}
    </button>
);

export default ControlButton;
