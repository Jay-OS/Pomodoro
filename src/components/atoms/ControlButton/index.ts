import ControlButton from './ControlButton';
import { controlButtonType, controlButtonSize } from './ControlButtonEnums';

export { controlButtonType, controlButtonSize };

export default ControlButton;
