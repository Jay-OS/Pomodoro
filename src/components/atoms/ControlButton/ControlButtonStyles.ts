import { StyleSheet } from 'aphrodite/no-important';

import colours from '../../../theme/styles/colours';
import { controlButtonType, controlButtonSize } from './ControlButtonEnums';

import { styleMapProperties } from '../../../theme/styles/objectMaps';

const controlButtonStyles: styleMapProperties = {
    controlButton: {
        borderRadius: '50%',
        border: 'none',
        padding: '0',
        transition: 'background-color 0.3s linear, color 0.3s linear',
        textAlign: 'center',
        ':hover': {
            backgroundColor: colours.BRAND_MAIN,
            color: 'white',
        },
    },
    [controlButtonType.PRIMARY]: {
        backgroundColor: colours.BRAND_MAIN,
        color: 'white',
    },
    [controlButtonType.SECONDARY]: {
        backgroundColor: colours.BRAND_SECONDARY,
        color: 'white',
    },
    [controlButtonType.REVERSED]: {
        backgroundColor: 'white',
        color: 'black',
    },
    [controlButtonSize.LARGE]: {
        height: '2.7rem',
        width: '2.7rem',
        fontSize: '1.6rem',
        lineHeight: '1.5rem',
    },
    [controlButtonSize.SMALL]: {
        height: '1.8rem',
        width: '1.8rem',
        fontSize: '1rem',
        lineHeight: '1rem',
    },
}

export default StyleSheet.create(controlButtonStyles);
