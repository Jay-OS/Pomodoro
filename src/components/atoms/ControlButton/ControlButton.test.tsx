import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import ControlButton, { ControlButtonProps } from './ControlButton';
import { controlButtonType, controlButtonSize } from './ControlButtonEnums';

import { mountTest, accessibilityChecks } from '../../../utils/testing';

describe('ControlButton . . .', () => {
    const testProps: ControlButtonProps = {
        id: 'button-id',
        title: 'Button title',
        onClick: () => {},
        disabled: false,
        'aria-label': 'Button label',
        buttonType: controlButtonType.PRIMARY,
        buttonSize: controlButtonSize.LARGE,
    };

    const renderComponent = (renderProps?: ControlButtonProps) => {
        const componentProps = renderProps ?? testProps;
        return <ControlButton {...componentProps}>Test</ControlButton>;
    };

    mountTest(renderComponent());
    accessibilityChecks(renderComponent());

    it('Matches snapshot', () => {
        const { container } = render(renderComponent());
        expect(container).toMatchSnapshot();
    });

    it('Calls onClick method when clicked', () => {
        const mockOnClick = jest.fn();
        render(renderComponent({ ...testProps, onClick: mockOnClick }));
        const button = screen.getByRole('button');

        expect(button).toBeVisible();
        expect(mockOnClick).not.toHaveBeenCalled();

        fireEvent.click(button);

        expect(mockOnClick).toHaveBeenCalledTimes(1);
    });
});
