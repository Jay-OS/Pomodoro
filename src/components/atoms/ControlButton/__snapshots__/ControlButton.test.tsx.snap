// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ControlButton . . . Matches snapshot 1`] = `
<div>
  <button
    aria-label="Button label"
    class="controlButton_ktbe5f-o_O-controlPrimary_1on53jx-o_O-controlLarge_zmmh0y"
    id="button-id"
    title="Button title"
  >
    Test
  </button>
</div>
`;
